### ISSUE TITLE
sql/pgwire: beekeeper studio gets unknown portal error while using portals

### ISSUE THREAD

**Issue by rafiss** [cockroachdb/cockroach#74187]
**Describe the problem**

Running beekeeper studio tests results in:
```
  ● CockroachDB Tests › Common Tests › stream tests

    Failed: "READ ERROR, query
        SELECT *
        FROM \"public\".\"streamtest\"


        order by \"id\" ASC


        : unknown portal \"C_1\""

      14 |   })
      15 |
    > 16 |   test.only("stream tests", async () => {
         |        ^
      17 |     await getUtil().streamTests()
      18 |   })
      19 |

      at Env.fit (../../node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:91:24)
      at runCommonTests (tests/integration/lib/db/clients/all.js:16:8)
      at Suite.<anonymous> (tests/integration/lib/db/clients/cockroach.spec.js:38:5)
      at Suite.<anonymous> (tests/integration/lib/db/clients/cockroach.spec.js:37:3)
```

@rathboma initially reported this issue and actually got a different error:
`multiple active result sets not supported` (from #40195)

**To Reproduce**

Install nodejs 14 and docker, then:
```
git clone git@github.com:beekeeper-studio/beekeeper-studio.git
cd beekeeper-studio
git checkout cockroach-stream-fix
yarn install
yarn workspace beekeeper-studio test:integration apps/studio/tests/integration/lib/db/clients/cockroach.spec.js
```

**Expected behavior**
No error.

**Additional data / screenshots**
I took a Wireshark capture and found the following capture: 
[beekeeper-cockroachdb.pcapng.zip](https://github.com/cockroachdb/cockroach/files/7759675/beekeeper-cockroachdb.pcapng.zip)

It seems that the client is sending an `ExecutePortal` message after receiving a `PortalSuspended` message from the server. (Which it shouldn't be doing.)

---

The fix for this might be on the CockroachDB side or on the beekeeper side.

**Comment by rafiss** [comment]
@rathboma I'd be interested to know if this same test works against Postgres. CockroachDB speaks the same wire protocol, so you should just be able to point it at Postgres and see what happens.

**Comment by rathboma** [comment]
Hey! Yes this same test runs for Postgres just fine as part of my test suite.

**Comment by rathboma** [comment]
Am I doing some incorrect sequencing do you think on the cursor?

I'm really out of my depth with database internals, but presumably `ExecutePortal` is the result of a `read`?

I think the way my implementation works is that it keeps calling `read` until a `read` returns 0 rows, then it marks itself as finished.

So my question is -  Is the error because the server says `I'm done`, and then Beekeeper tries to `read` one more time?

Maybe I should be listening to an event instead?

**Comment by rathboma** [comment]
For reference, here was the documentation I followed which suggested the `read` behavior that I follow:

https://node-postgres.com/api/cursor

**Comment by rathboma** [comment]
Similar issue on the node-postgres driver reported in May - https://github.com/brianc/node-postgres/issues/2546

**Comment by rafiss** [comment]
Don't worry if you don't know the database internals! That's why I'm here to help :)

Your research there was helpful and led me to a few things.

As you say, this works with Postgres, so this seems like a bug in our implementation. It's possible it might be a known difference, but I haven't yet looked to see if that's the case.

> I'm really out of my depth with database internals, but presumably ExecutePortal is the result of a read?

Yeah, exactly that's right. A portal is the name of the database-internal object that has the state of how far you've read when you're doing a cursor-based query like this.

> So my question is - Is the error because the server says I'm done, and then Beekeeper tries to read one more time?

That was what I thought for a little while, but the way Beekeeper is working is correct to me. It should be allowed to read one more time.

---

The problem here has to do with how CockroachcDB deals with portals in implicit transactions (i.e. not using BEGIN/COMMIT). We seem to have a behavior where we automatcally close the portal at the end of an implicit transaction, which is not how Postgres works.

So based on my testing, a workaround here is to wrap the queries you're doing with BEGIN (before all the reads), and COMMIT (after you try a read and get 0 rows). I don't know if this change would be feasible for Beekeeper though, but that's just one idea.

I'll try to see if the divergence from Postgres can be fixed.

**Comment by rafiss** [comment]
@rathboma I have merged a fix that should make things work as you intended. However, currently I'm not sure if it will be available in a 21.2 release, or if we'll have to wait until v22.1 comes out in the spring of 2022.

In the meantime, please let me know if that workaround works for you.

**Comment by rathboma** [comment]
Wow, this is amazing, thanks so much for working on this so quickly.
